### ISSUE TITLE
Add the new type "Set"

### ISSUE THREAD

**Issue by RadiantUwU** [godotengine/godot-proposals#7155]
### Describe the project you are working on

Logic sim plugin

### Describe the problem or limitation you are having in your project

Having to call two methods instead of one, when an optimization can be done so instead of using an array, checking if its in it and if not to add it. The set is used for pins that would be requested to update.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Instead, a set will be optimized to only hold the object only once, which can speed up finding it, removing or adding the object to the set.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```gdscript
var s:=Set()
s.add("test")
assert(s.has("test"))
#s.remove("foo") # doesnt exist, would throw an error
s.remove_if_exists("foo") # doesn't exist, wouldn't throw an error
for o in s:
   print(o) # "test" printed
s.add("foo")
var bar:=Set(["foo","bar"])
for o in s.intersection(bar):# or s & bar
    print(o) # "bar" printed
for o in s.union(bar):# or s | bar or s + bar
    print(o) # "test", "foo", "bar" are printed
for o in s.difference(bar):# or s - bar
    print(o) # "test" printed
for o in s.symmetrical_difference(bar):# or s ^ bar
    print(o) # "test", "bar" printed
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

A few lines, not really, a whole class should be implemented, probably extending from refcounted.

### Is there a reason why this should be core and not an add-on in the asset library?

Implementation in GDScript would be slower, a set would just hash the object and look for it, similar to how a dictionary would,but having a new variant type would solve all problems.

**Comment by dalexeev** [comment]
Related:

* #867
* #1756

Adding a new Variant type should be strongly justified, as it slows down all operations. Instead, it can be an object (RefCounted). Also you can implement this class in GDScript, there is no need to have it in the core.

**Comment by Calinou** [comment]
- Duplicate of https://github.com/godotengine/godot-proposals/issues/867.
